### DISCUSSION TITLE
Is there a way to not execute commands on initial startup?

### DISCUSSION THREAD

**Discussion by 9las** [i3/i3#4824] [Q&A] [answered]
I would like to know if there is a way to not execute commands on initial startup.

The reason for this is that I would like autorandr to restart i3 whenever a new monitor setup is detected. This is achieved by a postswitch which contains the command `i3-msg restart` which again results in that my background is reset and polybar is being relaunched due to the `exec_always` lines in my i3 config. At the same time I also want autorandr to run at boot to detect my monitor setup which I have achieved by writing `exec --no-startup-id autorandr --change` in the i3 config file.

The problem is though that the background and polybar are being reset twice at boot due to the autorandr postswitch. Is there a way to run commands when restarting i3 but not on initial startup? Or does someone know of a better solution to my problem?

I tried passing an environment variable to autorandr to tell the postswitch to not run the command `i3-msg restart` at boot by using an if statement. This made it so that the `exec_always` commands were only run once, but they were executed too soon after autorandr, meaning that polybar and the background were not set correctly. Thus it is important that polybar and the background are first set after autorandr has finished. Which is the reason for the postswitch.

These are my current scripts and configs where polybar and the background are set twice at boot:

Relevant lines from i3 config:
```
# detect monitor setup and run postswitch
exec --no-startup-id autorandr --change

# set wallpaper
exec_always --no-startup-id feh --bg-fill $HOME/Pictures/Wallpapers/i-HZT8VDH-O.jpg

# launch polybar
exec_always --no-startup-id $HOME/.config/polybar/launch.sh

```

postswitch:
```
#!/usr/bin/env bash

# Restart i3 inplace (perserves layout/session)
i3-msg restart
```

`$HOME/.config/polybar/launch.sh` script:
```
# Terminate already running bar instances
polybar-msg cmd quit

# Wait until the processes have been shut down
while pgrep -x polybar >/dev/null; do sleep 1; done

# Launch Polybar, using default config location ~/.config/polybar/config
for monitor_name in $(polybar --list-monitors | cut -d":" -f1); do
    MONITOR=$monitor_name polybar example & disown
done
```

**Comment by Airblader** [comment, marked as answer]
> I would like to know if there is a way to not execute commands on initial startup.

Not in configuration directive form. Here's a bunch of ideas, there's probably even more ways of doing it.

1. Using `set_from_resources`. Untested, so maybe needs tweaking, but you get the idea I think:

```
# Define a variable which generally doesn't exist in xrdb
set_from_resources $initialStart i3wm.initialStart true

# Populate the xrdb for the next time
# You might have to move this to a script, or just the echo part and use xrdb -merge /path/to/file
exec echo "i3wm.initialStart: false" | xrdb -merge

# Use this to bypass polybar the first time around
exec_always $initialStart || polybar
```

2. You could have a script launched through `exec` which monitors for restart events in the IPC and then executes those commands (again through the IPC). Since the script is started on the initial start, it wouldn't catch that event and not execute the commands that first time.

3. Hook into where you start the i3 process, and create some "marker" file. Then instead of the commands you execute a script from your config that checks for the existence of that file; if it exists, delete it and exit. If it does not, execute the commands.

I just quickly laid out some ideas for the question asked here. I didn't think about whether the entire thing could just be done differently.

**Comment by 9las** [comment]
Thanks for the ideas. I ended up doing the following instead:

I have this in my i3 config:
```
# auto configure monitors and use postscript to set wallpaper and launch polybar
exec_always --no-startup-id $HOME/.config/i3/scripts/auto-configure_monitors.sh
```

The script `auto-configure_monitors.sh` then tries to change the monitor configuration using `autorandr --change` which also activates the postswitch. If the configuration is already loaded only the postswitch script will get executed. Here is the content of the script `auto-configure_monitors.sh`:
```
#!/usr/bin/env bash

# Try to change configuration and handle if it is already loaded
if autorandr --change 2>&1 >/dev/null | grep "Config already loaded" >/dev/null; then

# Execute postswitch
    bash $HOME/.config/autorandr/postswitch

fi
```
Content of postswitch:
```
#!/usr/bin/env bash

# set wallpaper
feh --bg-fill $HOME/Pictures/Wallpapers/i-HZT8VDH-O.jpg

# launch polybar
$HOME/.config/polybar/launch.sh
```

Content of `launch.sh`:
```
#!/usr/bin/env bash

# Terminate already running bar instances
polybar-msg cmd quit

# Wait until the processes have been shut down
while pgrep -x polybar >/dev/null; do sleep 1; done

# Launch Polybar, using default config location ~/.config/polybar/config
for monitor_name in $(polybar --list-monitors | cut -d":" -f1); do
    MONITOR=$monitor_name polybar example &
done
```

I also have the following line in the polybar config:
```
monitor = ${env:MONITOR:}
```
